cip/CIP2015-09-16-public-type-system-type-annotation.adoc

```diff
@@ -0,0 +1,259 @@
+ * `LOCALDATETIME`
+ ** An instant capturing the date and the time, but not the time zone.
+ * `DATE`
+ ** An instant capturing the date, but not the time, nor the time zone.
+ * `LOCALDATE`
+ ** An instant capturing the date, but not the time, nor the time zone.
+ * `LOCALTIME`
+ ** An instant capturing the time of day, but not the date, nor the time zone.
+ * `DURATION`
+ ** A temporal amount. This captures the difference in time between two instants. It only captures the amount of time between two instants, it does not capture a start time and end time. A unit capturing the start time and end time would be a _Time Interval_ and is out of scope for this proposal.
+
+Note that NULL is not a type but a value that inhabits every type.
+
+==== Type Annotation
+To specify the type type of a term term in future changes to the Cypher grammar, this CIP proposes using the following syntax
+
+[source, ebnf]
+----
+	term :: type
+----
+
+==== Type Literal Use Outside of Annotations
+Type literals could be used in other production rules as well where this is considered more readable by future CIPs (e.g. a type test operator expr IS NUMBER).
+
+=== 3.2 Syntax
+
+[source, ebnf]
+----
+type = scalar type
+     | container type
+     ;
+
+primitive type 	= "ANY"
+                | "BOOLEAN"
+                | "INTEGER"
+                | "FLOAT"
+                | "IDENTITY"
+                | "MAP"
+                | "NODE"
+                | "STRING"
+                | "RELATIONSHIP"
+                | "PATH"
+                | "NUMBER"
+                | "DATETIME"
+                | "LOCALDATETIME"
+                | "DATE"
+                | "LOCALDATE"
+                | "LOCALTIME"
+                | "DURATION”
+                ;
+
+container type	= "LIST", "<", type, ">";
+
+type annotation = "::", type ;
+----
+
+=== Semantics
+This CIP only provides syntax for future CIPs, it does not directly change Cypher and therefore does not change semantics.
+
+The intended use of type annotations is that they express that the annotated term either has, evaluates to, or is coerced to a value of the annotated type.
+
+=== Interaction with existing features
+This CIP adds the following new keywords for all type names. It is expected that these type name keywords are only valid in specific contexts in the grammar (mostly in type annotations but also possibly in operators). The likelihood of conflict with existing (or future) production rules is therefore minimal.
+
+=== Alternatives
+As part of writing this CIP, many syntax alternatives have been considered:
+
+* Alternative name for the `STRING` type: `TEXT`, `UNICODE`, `LIST<CHAR>`
+* Alternative name for the `LIST` type: `ARRAY`, `COLLECTION`, `SEQUENCE`, `VECTOR`
+* Alternative syntax for type parameters: `LIST OF T`, `LIST[T]`, `LIST<T>`, `LIST T`
+* Alternative syntax for type annotations: `(STRING) expr`, `STRING expr`, `expr: `T`
+
+== What others do
```

Comment:
Like someone else has commented somewhere else - this section only deals with casting and coercing, and not around the type system per se. A discussion on the type system compared to our competitors would be interesting